server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -1013,6 +1040,11 @@ public void writeTo(StreamOutput out) throws IOException {
         PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), INDEX_MODE);
         PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> DataStreamLifecycle.fromXContent(p), LIFECYCLE);
         PARSER.declareBoolean(ConstructingObjectParser.optionalConstructorArg(), ROLLOVER_ON_WRITE_FIELD);
+        PARSER.declareObject(
+            ConstructingObjectParser.optionalConstructorArg(),
+            (p, c) -> DataStreamAutoShardingEvent.fromXContent(p),
+            AUTO_SHARDING_FIELD
+        );
```

Comment:
Would we want this behind a check for the feature flag, the same way we have a check for the failure store feature flag? If we did, we'll need to adjust the `args[n]` offsets above.

Reply:
We don't have a feature flag, and I wasn't planning on adding one. The setting to enable the functionality (only available in certain environments) should act as the flag to enable/disable the feature. 

Were you thinking we need a feature flag?

Reply:
If we never have a situation where the xcontent pieces get written out to state, then it's fine not to have one. If we do, then we'll need one to ensure we don't have an unrecoverable cluster due to serialization issues.